react/features/device-selection/components/DeviceSelectionDialog.js

```diff
@@ -515,70 +394,120 @@ class DeviceSelectionDialog extends Component {
-                && this.props.currentVideoTrack.getDeviceId())
-            || (availableDevices.videoInput[0]
-                && availableDevices.videoInput[0].deviceId)
-            || ''; // DeviceSelector expects a string for prop selectedDeviceId.
-
+        const { _availableDevices } = this.props;
         const configurations = [
             {
-                devices: availableDevices.videoInput,
+                devices: _availableDevices.videoInput,
                 hasPermission: this.props.hasVideoPermission,
                 icon: 'icon-camera',
                 isDisabled: this.props.disableDeviceChange,
                 key: 'videoInput',
                 label: 'settings.selectCamera',
-                onSelect: this._getAndSetVideoTrack,
-                selectedDeviceId: currentVideoId
+                onSelect: this._updateVideoInput,
+                selectedDeviceId: this.state.selectedVideoInputId
             },
             {
-                devices: availableDevices.audioInput,
+                devices: _availableDevices.audioInput,
                 hasPermission: this.props.hasAudioPermission,
                 icon: 'icon-microphone',
                 isDisabled: this.props.disableAudioInputChange
                     || this.props.disableDeviceChange,
                 key: 'audioInput',
                 label: 'settings.selectMic',
-                onSelect: this._getAndSetAudioTrack,
-                selectedDeviceId: currentAudioId
+                onSelect: this._updateAudioInput,
+                selectedDeviceId: this.state.selectedAudioInputId
             }
         ];
 
         if (!this.props.hideAudioOutputSelect) {
             configurations.push({
-                devices: availableDevices.audioOutput,
+                devices: _availableDevices.audioOutput,
                 hasPermission: this.props.hasAudioPermission
                     || this.props.hasVideoPermission,
                 icon: 'icon-volume',
                 isDisabled: this.props.disableDeviceChange,
                 key: 'audioOutput',
                 label: 'settings.selectAudioOutput',
-                onSelect: this._getAndSetAudioOutput,
-                selectedDeviceId: currentAudioOutId
+                onSelect: this._updateAudioOutput,
+                selectedDeviceId: this.state.selectedAudioOutputId
             });
         }
 
         return configurations.map(this._renderSelector);
     }
+
+    /**
+     * Callback invoked when a new audio input device has been selected. Updates
+     * the internal state of the user's selection as well as the audio track
+     * that should display in the preview.
+     *
+     * @param {string} deviceId - The id of the chosen audio input device.
+     * @private
+     * @returns {void}
+     */
+    _updateAudioInput(deviceId) {
+        this.setState({
+            selectedAudioInputId: deviceId
+        }, () => {
+            this._disposeAudioPreview()
```

Comment:
What do you think of the following architecture:
Right now DeviceSelectionDialog is taking care of all tracks (creation, disposal, etc), right?
Isn't it going to be better if you are just passing the device IDs to the preview component and handle the tracks there. I think that will simplify the whole process.

Even if you agree with me we can do this trough another PR because I don't think this is in the scope of this one. I just wanted to share the idea :)